src/main/scala/org/neo4j/spark/rdd/Neo4jRowRDD.scala

```diff
@@ -39,15 +38,17 @@ class Neo4jRowRDD(@transient sc: SparkContext, val query: String, val parameters
           }
           Row.fromSeq(builder.result())
         }
-        if (!result.hasNext) {
-          if (session.isOpen) session.close()
-          driver.close()
-        }
         res
       })
-    } finally {
-      if (session.isOpen) session.close()
-      driver.close()
+      new LazySessionIterator(it, driver, session)
+    } catch {
+      case e: Exception => {
+        if (session.isOpen) {
```

Comment:
null checks and try-catch for session closing 
that's why I wrapped this in a function that I could reuse in all the places !

Reply:
ok